nats.go

```diff
@@ -2494,30 +2501,41 @@ func (nc *Conn) processMsg(data []byte) {
 
 	// We have two modes of delivery. One is the channel, used by channel
 	// subscribers and syncSubscribers, the other is a linked list for async.
-	if sub.mch != nil {
+	if !ctrl {
+		if sub.mch != nil {
+			select {
+			case sub.mch <- m:
+			default:
+				goto slowConsumer
+			}
+		} else {
+			// Push onto the async pList
+			if sub.pHead == nil {
+				sub.pHead = m
+				sub.pTail = m
+				if sub.pCond != nil {
+					sub.pCond.Signal()
+				}
+			} else {
+				sub.pTail.next = m
+				sub.pTail = m
+			}
+		}
+	} else if jsi != nil && jsi.fch != nil {
+		// Process flow control and heartbeat messages
+		// automatically for JetStream Push consumers.
 		select {
-		case sub.mch <- m:
+		case jsi.fch <- m:
```

Comment:
In order to support ChanSubscribe, SyncSubscribe and Async Subscribers, a goroutine that expects signals via a channel is started to process the status messages outside of the message delivery.